testcases/kernel/syscalls/madvise/madvise08.c

```diff
@@ -0,0 +1,255 @@
+	tst_res(TINFO, "Temporary core pattern is '%s'", tmpcpattern);
+	SAFE_FILE_PRINTF(CORE_PATTERN, "%s", tmpcpattern);
+
+	tc.fmem = SAFE_MMAP(NULL,
+			    tc.fmemsize,
+			    PROT_READ | PROT_WRITE,
+			    MAP_ANONYMOUS | MAP_PRIVATE,
+			    -1,
+			    0);
+
+	/*Write a generated character sequence to the mapped memory,
+	  which we later look for in the core dump.*/
+	fmemc = (char *)tc.fmem;
+	*fmemc = 'x';
+	for (i = 0; i < tc.ycount; i++)
+		fmemc[i + 1] = 'y';
+	fmemc[++i] = 'z';
+}
+
+static void cleanup(void)
+{
+	if (tc.cpattern != NULL) {
+		if(FILE_PRINTF(CORE_PATTERN, "%s", tc.cpattern)) {
+			tst_brk(TBROK,
+				"COULD NOT RESTORE "CORE_PATTERN"! It used to be '%s'",
+				tc.cpattern);
+		}
+		free(tc.cpattern);
+	}
+
+	if (tc.fmem != NULL && munmap(tc.fmem, tc.fmemsize))
+		tst_res(TBROK | TERRNO, "munmap(tc.fmem)");
+	tc.fmem = NULL;
+
+	if (tc.dfd > 0 && close(tc.dfd))
+		tst_res(TBROK | TERRNO, "close(tc.dfd)");
+}
+
+static int find_sequence(int pid)
+{
+	char expectc = 'x';
+	ssize_t read, pos = 0;
+	char rbuf[1024];
+	int ycount = 0;
+	char dumpname[256];
+
+	snprintf(dumpname, 256, "dump-%d", pid);
+	tst_res(TINFO, "Dump file should be %s", dumpname);
+	if (access(dumpname, F_OK))
+		tst_brk(TBROK | TERRNO, "Dump file was not found.");
+
+	tc.dfd = SAFE_OPEN(dumpname, O_RDONLY);
+
+	read = SAFE_READ(0, tc.dfd, &rbuf, sizeof(rbuf));
+	while (read) {
+		switch (rbuf[pos]) {
+		case 'x':
+			ycount = 0;
+			expectc = 'y';
+			break;
+		case 'y':
+			if (expectc == 'y') {
+				ycount++;
+			} else {
+				expectc = 'x';
+				break;
+			}
+
+			if (ycount == tc.ycount)
+				expectc = 'z';
+			break;
+		case 'z':
+			if (expectc == 'z')
+				return 1;
+		default:
+			expectc = 'x';
+		}
+		if (++pos >= read) {
+			read = SAFE_READ(0, tc.dfd, &rbuf, sizeof(rbuf));
+			pos = 0;
+		}
+	}
+
+	SAFE_CLOSE(tc.dfd);
+	tc.dfd = 0;
+	return 0;
+}
+
+static pid_t run_child(int advice)
+{
+	int status;
+	pid_t pid;
+	char *advstr =
+		advice == MADV_DONTDUMP ? "MADV_DONTDUMP" : "MADV_DODUMP";
+
+	pid = SAFE_FORK();
+	if (pid == 0) {
+		if (madvise(tc.fmem, tc.fmemsize, advice) == -1) {
+			tst_res(TFAIL | TERRNO,
+				"madvise(%p, %lu, %s) = -1",
+				tc.fmem,
+				tc.fmemsize,
+				advstr);
+		} else {
+			abort();
+		}
+	}
+
+	SAFE_WAITPID(pid, &status, 0);
+	if (WIFSIGNALED(status) && !WCOREDUMP(status)) {
+		tst_res(TCONF, "No core dump produced after abort(), see core(5).");
+		pid = 0;
+	} else if (!WIFSIGNALED(status)) {
+		pid = 0;
+	}
+
+	return pid;
+}
+
+static void run(void)
+{
+	pid_t pid;
+
+	pid = run_child(MADV_DONTDUMP);
+	if (pid && find_sequence(pid))
+		tst_res(TFAIL, "Found sequence in dump when MADV_DONTDUMP set");
+	else if (pid)
+		tst_res(TPASS, "madvise(..., MADV_DONTDUMP");
+
+	pid = run_child(MADV_DODUMP);
+	if (pid && find_sequence(pid))
+		tst_res(TPASS, "madvise(..., MADV_DODUMP)");
+	else if (pid)
+		tst_res(TFAIL, "No sequence in dump after MADV_DODUMP.");
+}
```

Comment:
Yes, I guess I need to either introduce a program argument to specify the test or split it into two executables. I think a program argument would be best.